[35기 2사리빙 홍현진] CategoryView 작성

hjz1010 opened the pull request:
## :: 최근 작업 주제 (하나 이상의 주제를 선택해주세요.)
- [x] 기능 추가
- [x] 리뷰 반영
- [ ] 리팩토링
- [ ] 버그 수정
- [ ] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 사용자가 대분류를 클릭하면 대분류에 속하는 모든 제품의 목록을 표출하는 기능 구현 
- >  카테고리(대분류or소분류)를 클릭하면 그 카테고리에 속하는 모든 제품의 목록을 표출하는 기능 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. client에서 path parameter로 category_id를 담아서 요청하면
2. 요청받은 category_id에 해당하는 모든 제품을 DB에서 찾아서
3. 각 제품의 id, 썸네일 이미지 URL, brand 이름, 제품 이름 + 색상, 가격 을 반환해준다!


1. 대분류와 소분류의 경우를 다른 view로 나누지 않고 하나로 통일하기 위해 query parameter 로 변경
2. query parameter 에서 카테고리 id를 추출하여 요청하는 제품의 정보를 반환!!

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- get 요청시 request.body에 요청 정보를 담아서 보내는 것이 아니라
- URL을 통해서 category_id 등을 받고 
- 이를 get메소드의 인자로 받아서 데이터 가공에 사용하는 방법을 배웠습니다.
 
- Path Parameter가 아니라 Query Parameter로 요청 정보를 받는 방법도 터득!!!

<br />

## :: 기타 질문 및 특이 사항
- View 이름도 적합하게 변경했는데 브랜치도 바꾸면 PR이 쌓이지 않아서 그냥 요기 브랜치에 push 했습니다.
- 앞으로도 비슷한 경우가 있다면 이렇게 하는게 좋은가요, 매치되는 브랜치를 새로 생성하고 이 브랜치는 삭제하는 것이 좋은가요?

hjz1010 committed:
ADD: 대분류 요청시 해당하는 제품의 목록을 반환하는 CategoryView 작성

hjz1010 committed:
csv 파일 -> DB

hjz1010 committed:
MODIFY: 카테고리 선택시 해당 제품의 리스트를 반환하는 ProductListView로 수정

Kang-kyunghun commented on products/urls.py:
```diff
@@ -0,0 +1,10 @@
+
+urlpatterns = [
+    path('', ProductListView.as_view()),
+    # 127.0.0.1:8000/products?category_id=2
+    # 127.0.0.1:8000/products?sub_category_id=7    
```
불필요한 주석 삭제

Kang-kyunghun commented on db_uploader.py:
```diff
@@ -0,0 +1,72 @@
+import csv
```
gitignore에 추가해 주세요.

hjz1010 replied:
추가하고 업로드 됐던 파일 삭제했습니다:)

Kang-kyunghun commented on products.txt:
```diff
@@ -0,0 +1,31 @@
+id	category	sub_category	brand_name	furniture_korean_name	furniture_english_name	color_korean_name	color_english_name	price(8)	thumbnail_image_url(9)	main_image_url	detail_image_url	is_new	
```
gitignore에 추가해 주세요.

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,28 @@
+        category_id     = request.GET.get('category_id', None)
+        sub_category_id = request.GET.get('sub_category_id', None)
+
+        products = Product.objects.filter(sub_category_id = sub_category_id) or Product.objects.filter(sub_category__category_id = category_id) 
```
filter를 구현하기 위해 Q객체를 조사해 보면 좋을 것 같습니다.

hjz1010 replied:
Q객체를 이용하면서 구현하려던 방향으로 기능이 작동하도록 수정해보았습니다!

hjz1010 committed:
ADD: ProductListView에 paginator 추가

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,28 @@
+            })
+        if product_list:
+            return JsonResponse({'message': 'SUCCESS', 'product_list': product_list}, status=200)
+        else:
+            return JsonResponse({'message': 'INVALID_CATEGORY'}, status=400)
```
product_list가 빈 list이면 빈 list 그대로 보내셔도 좋을 것 같습니다.

hjz1010 replied:
수정했습니다!

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,28 @@
+from django.http  import JsonResponse
+from django.views import View
+
+from .models import Category, Furniture, ProductImage, SubCategory, Product, Brand, Color 
```
사용하지 않는 class 들은 정리 한 번 해주세요.

hjz1010 replied:
다 해두는게 좋은게 아니었군요.. ㅎㅎ

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,3 +1,28 @@
+                'brandName'  : product.furniture.brand.name,
+                'productName': product.furniture.korean_name + '_' + product.color.korean_name,
+                'price'      : product.price
+            })
```
list comprehension으로 바꿔 보세요.

hjz1010 replied:
컴프리헨션이 가독성이 훨씬 좋네요😁

Kang-kyunghun commented on products.txt:
```diff
@@ -0,0 +1,55 @@
+id	category	sub_category	brand_name	furniture_korean_name	furniture_english_name	color_korean_name	color_english_name	price	thumbnail_image_url	main_image_url	detail_image_url	is_new
```
gitignore에 추가해 주세요.

hjz1010 committed:
MODIFY: 수정요청 사항 반영 1차(Q객체는 아직)

hjz1010 committed:
컨벤션 수정

hjz1010 committed:
컨벤션 수정

hjz1010 committed:
MODIFY: 수정요청 사항 반영 완료

hjz1010 committed:
.gitignore 추가된 파일, 업로드된 것 삭제

hjz1010 committed:
ProductDetailView merge

hjz1010 committed:
MODIFY: Q 객체 사용방법 수정

hjz1010 committed:
MODIFY: ProductListView 예외처리 추가

hjz1010 committed:
MODIFY: ProductListView 수정요청 반영 및 정렬 기능 구현
- pagination, offset+limit로 변경
- 동일한 if문 반복 사용되는 것 수정

- 정렬(id순, 높은가격순, 낮은가격순)

hjz1010 committed:
MODIFY: ProductListView 수정
- sort 조건에 신상품순 추가
- INVALID_OFFSET 예외처리 추가

hjz1010 committed:
MODIFY: INVALID_OFFSET 예외처리, 불필요한 것 같아서 삭제

hjz1010 committed:
컨벤션 수정

hjz1010 committed:
이 브랜치에 해당하지 않는 변경사항 삭제

Kang-kyunghun commented on products/views.py:
```diff
@@ -1,8 +1,63 @@
+                sub_category_q &= Q(category=sub_category.category)
+                product_q      &= Q(sub_category = sub_category)
+            
+            count = len(Product.objects.filter(product_q))
```
```suggestion
            count = Product.objects.filter(product_q).count()
```

hjz1010 committed:
MODIFY: 수정요청 반영
- sort_type 숫자가 아닌 string값으로 명확하게 받아오기
- count 구할 때 전부 필터링 후에 쿼리셋의 길이를 구하지 말고 .count()로 효율적으로
- sub_category_list 제공은 API 따로 만들기

hjz1010 committed:
conflict 해결

hjz1010 committed:
sort_type 받아올때 int로 변환한거 삭제

Kang-kyunghun commented:
LGTM
